cmd/repo_list.go

```diff
@@ -16,8 +17,10 @@ func aptlyRepoList(cmd *commander.Command, args []string) error {
 	}
 
 	raw := cmd.Flag.Lookup("raw").Value.Get().(bool)
+	jsonFlag := cmd.Flag.Lookup("json").Value.Get().(bool)
 
 	repos := make([]string, context.CollectionFactory().LocalRepoCollection().Len())
+	jsonRepos := make([]*deb.LocalRepo, context.CollectionFactory().LocalRepoCollection().Len())
```

Comment:
Are we sure that collecting full `deb.LocalRepo` objects rather than just strings (as in the `repos` array) is not going to use too much memory?

If yes, then why not *always* collect all repos in an array and do the formatting later, either with `fmt.Sprintf` or with `json.Marshal` if the `-json` flag is set?

At the moment, the same function handles both txt and json output: but it's so different, because of different variables or switches, that I wonder if it doesn't make more sense to have separate functions (if the above method doesn't work).

Reply:
Thanks for the review!

I'm not sure about the memory issue. Do you know if any statistics are available showing the maximum number of repo/mirror/snapshot/publish supported and/or expected?

As for the second statement, the primary reason that I chose not to do that is because I'm sure there are various scripts in the wild that parse the output of aptly and I didn't want to break user's existing scripts. I'm happy to refactor it into a separate function though, if you feel that would be more appropriate.

Reply:
I guess a good metric is to see what portion of a function are common between "txt" and "json" outputs and make a judgement if it makes sense to split them in separate methods. I'll leave this judgement to you, of course.

And sure, we should keep backward compatibility, to avoid breaking existing workflows.

Reply:
Hmm, I'm curious what scale of repos you have in your internal aptly deployment. I just created 10k test repos on my dev system and this was the result:

```sh
> time aptly repo list -json
[
  {
    "Name": "test-repo-00001",
    "Comment": "",
    "DefaultDistribution": "",
    "DefaultComponent": "main"
  },
 # SNIP
  {
    "Name": "test-repo-09999",
    "Comment": "Comment on repo 09999",
    "DefaultDistribution": "bionic",
    "DefaultComponent": "main"
  },
  {
    "Name": "test-repo-10000",
    "Comment": "Comment on repo 10000",
    "DefaultDistribution": "bionic",
    "DefaultComponent": "main"
  }
]

real    0m0.764s
user    0m0.258s
sys     0m0.132s

Reply:
Actually just about 3000. I may be misremembering the slowness: I can't stop production to check again though atm.